gips/data/sentinel2/sentinel2.py

```diff
@@ -910,6 +917,88 @@ def ref_geoimage(self, asset_type, sensor):
             sr_image[c] = (rad_rev_img[c] - Lu) / TLdS
         self._product_images['ref'] = sr_image
 
+
+    def _band_filename(self, band, zipfile, asset_style):
+        for name in zipfile.namelist():
+            match = re.match(self.Asset._asset_styles[asset_style]['raster-re'], name)
+            if match and match.group('band') == band:
+                return name
+        raise IOError("{} does not contain raster band {}".format(
+                zipfile.filename, band))
+
+
+    def fmask_geoimage(self, asset_type):
+        """Generate cloud mask.
+
+        Uses python implementation of cfmask. Builds a VRT of all the necessary
+        bands in the proper order, an angles image is created using the supplied
+        metadata, and then the two are put through the fmask algorithm.
+        """
+        self._time_report('Generating cloud mask')
+
+        safe_zip = zipfile.ZipFile(self.assets[asset_type].filename, 'r')
+        safe_zip.extractall(self._temp_proc_dir)
+
+        asset_style = self.assets[asset_type].style
+        bands = []
+        for i in range(1, 9):
+            band = self._band_filename('0' + str(i), safe_zip, asset_style)
+            bands.append("{}/{}".format(self._temp_proc_dir, band))
+
+        band = self._band_filename('8A', safe_zip, asset_style)
+        bands.append("{}/{}".format(self._temp_proc_dir, band))
+
+        band = self._band_filename('09', safe_zip, asset_style)
+        bands.append("{}/{}".format(self._temp_proc_dir, band))
+
+        for i in range(3):
+            band = self._band_filename('1' + str(i), safe_zip, asset_style)
+            bands.append("{}/{}".format(self._temp_proc_dir, band))
+
+        gdalbuildvrt_args = [
+            "gdalbuildvrt",
+            "-resolution", "user",
+            "-tr", "20", "20",
+            "-separate",
+            "%s/allbands.vrt" % self._temp_proc_dir,
+        ] + bands
+        rc = subprocess.call(gdalbuildvrt_args)
+        if rc != 0:
+            raise IOError("Expected gdalbuildvrt exit status 0, got {}".format(rc))
+
+        metadata_xml = None
+        for name in safe_zip.namelist():
+            if re.match(self.Asset._asset_styles[asset_style]['tile-md-re'], name):
+                metadata_xml = "{}/{}".format(self._temp_proc_dir, name)
+        if not metadata_xml:
+            raise IOError("{} does not contain expected metadatafile.".format(safe_zip.filename))
+        rc = subprocess.call(
+            [
+                "fmask_sentinel2makeAnglesImage.py",
+                "-i", metadata_xml,
+                "-o", "%s/angles.img" % self._temp_proc_dir,
+            ]
+        )
+        if rc != 0:
+            raise IOError("Expected fmask_sentinel2makeAnglesImage.py exit status 0, got {}".format(rc))
+
+        rc = subprocess.call(
+            [
+                "fmask_sentinel2Stacked.py",
```

Comment:
the Word of @naterubin is that this will be added to gips install process at some point